File: test/errors.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { AssignmentActivityError } from "../src/activity.js";
import { errorRecord, jsonRequested, retainedAssignmentFailure } from "../src/errors.js";
import { ProcessIdentityUnavailableError } from "../src/process.js";

test("structured failures expose stable automation categories", () => {
  assert.deepEqual(errorRecord(new Error("Workspace has uncommitted changes.")), {
    status: "error",
    code: "WORKSPACE_DIRTY",
    message: "Workspace has uncommitted changes.",
    retryable: false,
  });
  assert.equal(errorRecord(new Error("Could not allocate an available port")).code, "PORT_UNAVAILABLE");
  assert.equal(errorRecord(new Error("Dependency installation failed")).code, "DEPENDENCY_PREPARATION_FAILED");
  assert.equal(errorRecord(new Error("Cannot read /repo/.rukrc.json: malformed JSON")).code, "INVALID_ARGUMENT");
  assert.deepEqual(
    errorRecord(new Error("bun 1.3.14 or newer is required for Ruk's shared dependency backend")),
    {
      status: "error",
      code: "DEPENDENCY_PREPARATION_FAILED",
      message: "bun 1.3.14 or newer is required for Ruk's shared dependency backend",
      retryable: true,
    },
  );
  assert.deepEqual(errorRecord(new ProcessIdentityUnavailableError(42)), {
    status: "error",
    code: "RESOURCE_BUSY",
    message: "Process 42 could not be identified, so its workspace cannot be released safely",
    retryable: true,
  });
  const retained = retainedAssignmentFailure(
    "00000000-0000-4000-8000-000000000000",
    "/workspace",
    "2026-08-15T00:00:00.000Z",
    new AggregateError([new ProcessIdentityUnavailableError(42)], "cleanup failed"),
  );
  assert.ok(retained);
  assert.deepEqual(errorRecord(retained), {
    status: "error",
    code: "RESOURCE_BUSY",
    message: "Assignment 00000000-0000-4000-8000-000000000000 retained at /workspace: cleanup failed",
    retryable: true,
    assignmentId: "00000000-0000-4000-8000-000000000000",
    path: "/workspace",
    expiresAt: "2026-08-15T00:00:00.000Z",
    recovery: "ruk release 00000000-0000-4000-8000-000000000000",
  });
  assert.equal(retainedAssignmentFailure("id", "/workspace", "expiry", new Error("failed")), null);
  assert.equal(
    errorRecord(new AggregateError([
      new Error("heartbeat failed"),
      new ProcessIdentityUnavailableError(42),
    ], "activity and cleanup failed")).code,
    "RESOURCE_BUSY",
  );
  assert.deepEqual(errorRecord(new Error("Could not enumerate POSIX processes: unavailable")), {
    status: "error",
    code: "RESOURCE_BUSY",
    message: "Could not enumerate POSIX processes: unavailable",
    retryable: true,
  });
  assert.deepEqual(errorRecord(new AssignmentActivityError("assignment-id", new Error("EPERM"))), {
    status: "error",
    code: "RESOURCE_BUSY",
    message: "Assignment assignment-id activity renewal failed: EPERM",
    retryable: true,
  });
  assert.equal(errorRecord(new Error("unexpected")).code, "OPERATION_FAILED");
  assert.equal(jsonRequested(["status", "--json"]), true);
  assert.equal(jsonRequested(["exec", "branch", "--", "tool", "--json"]), false);
  assert.equal(jsonRequested(["run", "tool", "--json"]), false);
});

File: internal/cli/errors_test.go
package cli

import (
	"errors"
	"strings"
	"testing"

	"github.com/xenoviz/ruk/internal/dependencies"
	"github.com/xenoviz/ruk/internal/lock"
	processpkg "github.com/xenoviz/ruk/internal/process"
)

func TestClassifyErrorPreservesStableCategories(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name      string
		err       error
		code      ErrorCode
		retryable bool
	}{
		{name: "dirty workspace", err: errors.New("Workspace has uncommitted changes."), code: WorkspaceDirtyCode},
		{name: "invalid config", err: errors.New("Cannot read /repo/.rukrc.json: malformed JSON"), code: InvalidArgumentCode},
		{name: "unknown option", err: errors.New("Unknown option --force"), code: InvalidArgumentCode},
		{name: "dependency message", err: errors.New("Dependency installation failed"), code: DependencyPreparationCode, retryable: true},
		{name: "typed dependency", err: &dependencies.DependencyPreparationError{Cause: errors.New("installer exited")}, code: DependencyPreparationCode, retryable: true},
		{name: "port", err: errors.New("Could not allocate an available port"), code: PortUnavailableCode, retryable: true},
		{name: "assignment conflict", err: errors.New("Assignment a does not exist"), code: AssignmentConflictCode},
		{name: "lock timeout", err: &lock.TimeoutError{Path: "/tmp/lock"}, code: ResourceBusyCode, retryable: true},
		{name: "process identity", err: &processpkg.IdentityUnavailableError{PID: 42, Cause: errors.New("identity unavailable")}, code: ResourceBusyCode, retryable: true},
		{name: "process enumeration", err: errors.New("Could not enumerate POSIX processes: unavailable"), code: ResourceBusyCode, retryable: true},
		{name: "git", err: errors.New("git fetch origin main: remote rejected"), code: GitOperationFailedCode},
		{name: "generic", err: errors.New("unexpected"), code: OperationFailedCode},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()
			record := ClassifyError(test.err)
			if record.Status != "error" {
				t.Fatalf("status = %q, want error", record.Status)
			}
			if record.Code != test.code {
				t.Fatalf("code = %q, want %q", record.Code, test.code)
			}
			if record.Retryable != test.retryable {
				t.Fatalf("retryable = %t, want %t", record.Retryable, test.retryable)
			}
		})
	}
}

func TestClassifyErrorPreservesRecoveryMetadata(t *testing.T) {
	t.Parallel()

	retained := RetainedAssignmentFailure(
		"assignment-1",
		"/workspace",
		"2026-08-15T00:00:00.000Z",
		&processpkg.IdentityUnavailableError{PID: 42},
	)
	if retained == nil {
		t.Fatal("RetainedAssignmentFailure returned nil")
	}
	record := ClassifyError(retained)
	if record.Code != ResourceBusyCode || !record.Retryable {
		t.Fatalf("record = %#v, want retryable resource busy", record)
	}
	if record.AssignmentID != "assignment-1" || record.Path != "/workspace" || record.ExpiresAt != "2026-08-15T00:00:00.000Z" {
		t.Fatalf("recovery metadata = %#v", record)
	}
	if record.Recovery != "ruk release assignment-1" {
		t.Fatalf("recovery = %q", record.Recovery)
	}
	if RetainedAssignmentFailure("id", "/workspace", "expiry", errors.New("ordinary failure")) != nil {
		t.Fatal("ordinary failure was incorrectly retained")
	}

	shared := ClassifyError(NewSharedCheckoutError(2))
	if shared.ActiveAssignments != 2 || shared.Recovery != "ruk acquire <branch>" {
		t.Fatalf("shared checkout metadata = %#v", shared)
	}
}

func TestClassifyErrorSearchesJoinedCauses(t *testing.T) {
	t.Parallel()

	err := errors.Join(
		errors.New("heartbeat failed"),
		NewAssignmentActivityError("assignment-1", errors.New("EPERM")),
	)
	record := ClassifyError(err)
	if record.Code != ResourceBusyCode || !record.Retryable {
		t.Fatalf("record = %#v, want retryable resource busy", record)
	}

	if got := ClassifyError(errors.Join(
		&dependencies.DependencyPreparationError{Cause: errors.New("dependency preparation aborted")},
		&processpkg.IdentityUnavailableError{PID: 42},
	)); got.Code != DependencyPreparationCode {
		t.Fatalf("joined typed dependency code = %q, want %q", got.Code, DependencyPreparationCode)
	}
}

func TestErrorFormattingAndJSONRequested(t *testing.T) {
	t.Parallel()

	err := errors.New("Workspace has uncommitted changes.")
	json := FormatJSONError(err)
	if !strings.HasSuffix(json, "\n") || !strings.Contains(json, `"code":"WORKSPACE_DIRTY"`) {
		t.Fatalf("JSON = %q", json)
	}
	if got := FormatHumanError(err); got != "ruk: Workspace has uncommitted changes.\n" {
		t.Fatalf("human = %q", got)
	}

	tests := []struct {
		args []string
		want bool
	}{
		{args: []string{"status", "--json"}, want: true},
		{args: []string{"exec", "branch", "--", "tool", "--json"}, want: false},
		{args: []string{"run", "tool", "--json"}, want: false},
		{args: []string{"list", "--json", "--", "ignored"}, want: true},
		{args: []string{"list", "--", "--json"}, want: false},
	}
	for _, test := range tests {
		if got := JSONRequested(test.args); got != test.want {
			t.Errorf("JSONRequested(%q) = %t, want %t", test.args, got, test.want)
		}
	}
}
